frontend/src/components/PromptInput.tsx

```diff
@@ -1,67 +1,242 @@
+  // from slerp to linear (e.g., "Switched to linear interpolation: Slerp requires exactly 2 prompts")
+  useEffect(() => {
+    if (prompts.length > 2 && interpolationMethod === "slerp") {
+      onInterpolationMethodChange?.("linear");
+    }
+  }, [prompts.length, interpolationMethod, onInterpolationMethodChange]);
 
-    setIsProcessing(true);
+  const handlePromptTextChange = (index: number, text: string) => {
+    const newPrompts = [...prompts];
+    newPrompts[index] = { ...newPrompts[index], text };
+    onPromptsChange?.(newPrompts);
+  };
 
-    // Send the prompt update via data channel
-    if (onPromptSubmit) {
-      onPromptSubmit(currentPrompt.trim());
+  const handleWeightChange = (index: number, weight: number) => {
+    const newPrompts = [...prompts];
+    newPrompts[index] = { ...newPrompts[index], weight };
+    onPromptsChange?.(newPrompts);
+  };
+
+  const handleAddPrompt = () => {
+    if (prompts.length < 4) {
+      onPromptsChange?.([...prompts, { text: "", weight: 100 }]);
     }
+  };
+
+  const handleRemovePrompt = (index: number) => {
+    if (prompts.length > 1) {
+      const newPrompts = prompts.filter((_, i) => i !== index);
+      onPromptsChange?.(newPrompts);
+    }
+  };
+
+  const handleSubmit = () => {
+    const validPrompts = prompts.filter(p => p.text.trim());
+    if (!validPrompts.length) return;
+
+    setIsProcessing(true);
+    onPromptsSubmit?.(validPrompts);
 
-    // Reset processing state after a short delay
     setTimeout(() => {
       setIsProcessing(false);
     }, 1000);
   };
 
-  const handleKeyPress = (e: React.KeyboardEvent) => {
-    if (e.key === "Enter") {
+  const handleKeyDown = (e: React.KeyboardEvent) => {
+    if (e.key === "Enter" && !e.shiftKey) {
+      e.preventDefault();
       handleSubmit();
     }
   };
 
-  return (
-    <div
-      className={`flex items-center bg-card border border-border rounded-full px-4 py-3 gap-3 ${className}`}
-    >
-      <Input
-        placeholder="blooming flowers"
-        value={currentPrompt}
-        onChange={e => onPromptChange?.(e.target.value)}
-        onKeyPress={handleKeyPress}
-        disabled={disabled}
-        className="flex-1 bg-transparent border-0 text-card-foreground placeholder:text-muted-foreground focus-visible:ring-0 focus-visible:ring-offset-0 p-0 disabled:opacity-50 disabled:cursor-not-allowed"
-      />
-      <Button
-        onClick={handleSubmit}
-        disabled={disabled || !currentPrompt.trim() || isProcessing}
-        size="sm"
-        className="rounded-full w-8 h-8 p-0 bg-black hover:bg-gray-800 text-white disabled:opacity-50 disabled:cursor-not-allowed"
+  // Calculate normalized weights for display
+  const totalWeight = prompts.reduce((sum, p) => sum + p.weight, 0);
+  const normalizedWeights = prompts.map(
+    p => (totalWeight > 0 ? (p.weight / totalWeight) * 100 : 0)
+  );
+
+  const isSinglePrompt = prompts.length === 1;
+
+  // Render a single prompt field with expandable textarea
```

Comment:
I see that by default we use Input and then switch to Textarea on focus.

IMO the visual jump from Input -> Textarea is a bit jarring. See below:

https://github.com/user-attachments/assets/08d0ecb4-6910-4369-8629-2181a028b9c5

Could we instead always use Textarea but show a smaller version with preview text and then on focus expand it to show the full text?

Also refer to screenshot of wireframes I have in mind in other comment.